### ISSUE TITLE
Automatic Export error (with Zotero Standalone)

### ISSUE THREAD

**Issue by gracile-fr** [retorquere/zotero-better-bibtex#537]
I had an automatic export set for a collection, but it has failed : ID: 3RNZT5KX

**Comment by retorquere** [comment]
I see only one error in the log:

> Translator with ID 59cce211-9d77-4cdd-876d-6229ea20367f already loaded from "Bibliotheque et Archives Nationales du Quebec.js"

could it be you have a duplicate translator with that ID? BBT tries to reload the translators to install the bundled translators and it might be failing unceremoniously when it encounters this error.

**Comment by gracile-fr** [comment]
Yes I've seen this error but it's old and unrelated to BBT I think.

**Comment by gracile-fr** [comment]
But you're right this error is caused by a duplicate translator with the same ID (actually the same translator but with diacritics in its filename, something that Zotero has changed). I've deleted the litigious file. Still, auto export is not working, ID: R2V794XC.

**Comment by retorquere** [comment]
At least now there's a different error. Is extended debugging enabled? Because I see nothing from BBT in the debug log. If it wasn't enabled, please enabled it and send a new debug log; if it was, it's an early startup error. I hope it's the former -- early startup  errors are a pain in the ass to diagnose under windows, certainly for Standalone (this is standalone right?)

**Comment by gracile-fr** [comment]
What about these two reports: WPZMH886 and DRRCF4KW ?
Yes, it's Standalone.
(also trying to create a new export for these tests, I've some encoding problem, but let's see that later)

**Comment by retorquere** [comment]
In both WPZMH886 and DRRCF4KW I see 

> parent of candidate path 'C:\\Users\\Gr�goire\\Documents\\Th�se\\XeLaTeX\\@chap1.bib' exists but is not writable

This would mean that (despite what the message says) either the parent directory does not exist, or it is not writable.

I've checked in a change that distinguishes between these two error conditions. I'm a little worried about those unicode missing marks in that string -- the path is just what I've been handed the Zotero translator code. I do see elsewhere that BBT can access the correctly spelled folder (`Grégoire` and `Thèse`). But these paths will be in the autoexport config of BBT, so this will always fail (I think) unless these paths are fixed. The easiest way to do that is to delete the autoexport and capture it again.

**Comment by gracile-fr** [comment]
Indeed, I can delete the autoexport and capture it again, but that's not really a solution :-)
What surprises me is that it used to work correctly before, even with diacritics in the path. And I've just tested it again with a slightly shorter path (still containing accented chars): "C:\Users\Grégoire\Desktop\test.bib", then with the same path and it seems to work (except the previous litigious autoexport)…!?
ID: 6RC44WGT ; 6HJM4TNV

**Comment by retorquere** [comment]
6RC44WGT and 6HJM4TNV still have the same error report. Once the wrong paths are in the autoexport table there's no question on while they fail -- the question is how those broken paths got in there. In these reports, did you remove and re-add the auto-exports, or are these still the broken auto-exports you also had when you created  WPZMH886?

**Comment by gracile-fr** [comment]
No I created these auto-exports for testing purpose, and immediately add/delete items from the collection to test the auto-export.

**Comment by gracile-fr** [comment]
I've deleted all my auto-export, close Standalone, launch it again. If I create an auto-export, play with it, etc., it works well. The path is correct in the bbt prefs. But if if close and start Standalone again, the path in the prefs is changed with the accented chars replaced by the unicode missing marks…

**Comment by retorquere** [comment]
Wot?

Can you go into the Zotero data directory (Preferences/Advanced/Files and Folders and click "Show Data Directory") and drop `db.json` file you find in the better-bibtex directory into https://dbinbox.com/retorquere?

**Comment by gracile-fr** [comment]
Done. There's also db.json.1, db.json.2, db.json.3, db.json.4 and db.json.5 in the dir.

**Comment by retorquere** [comment]
Those are backups, so that's OK. Could you try https://github.com/retorquere/zotero-better-bibtex/issues/537#issuecomment-233906169 again with [this version](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.68-circle-2618.xpi)?

**Comment by retorquere** [comment]
Nothing in that log that seems relevant to BBT. There is an error in the log, but it seems unrelated to BBT -- please try [this version](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.68-circle-2619.xpi) which works around what appears to be a Zotero bug to see of the log now clears.

Could you get me a new copy of db.json on dbinbox, and a new error log with the version on the link?

**Comment by gracile-fr** [comment]
Installing the first custom version (i.e. zotero-better-bibtex-1.6.68-circle-2618) deleted all db.json except db.json.5
ID, anyway: 4RS5EWQ3

**Comment by retorquere** [comment]
4RS5EWQ3 has the same errors, so the workaround didn't do anything. That means that error must be before BBT initializes (I don't see the error, so I can't verify). Can you disable BBT, restart ZSA, and call up the error log from the Zotero advanced prefs to see if you see something like 

> [JavaScript Error: "TypeError: str.split is not a function" {file: "chrome://zotero/content/xpcom/utilities.js" line: 527}]

I have *no* idea how the db.json files disappeared. I'm going over my code but I just don't see a place that removes them -- specifically db.json.1-4

After restarting, and adding an auto-export, restarting again, there really must be a db.json file. Can you get me a copy of that?

**Comment by gracile-fr** [comment]
I can't see the "str.split is not a function" error in the zotero debug output…

**Comment by gracile-fr** [comment]
There's no db.json except db.json.5… I tried to reinstall the latest official version (1.6.68), create an auto-export etc, a db.json was recreated. But once I've updated to the latest custom version, it was removed.

**Comment by retorquere** [comment]
Argh! I take it you do see them when bbt is enabled? Wait, did you have
debug logging on? In the zotero advanced prefs, tick "enable after
restart", restart, and see what's in the log then.

**Comment by gracile-fr** [comment]
"enable after restart" was already checked in Z advanced prefs

**Comment by retorquere** [comment]
Do you have an idea when it is removed? Before restart, after restart? I'm looking at the diff between the released version and the prereleases you've tried, and there's really nothing in those 5 lines that touches the backups. In the logs, I can see the backups being created.

**Comment by gracile-fr** [comment]
It is created immediately after starting Zotero Standalone with 1.6.68 :  "db1.6.68.json" - 8.8 kB
It is modified after the installation of 1.6.68-circle-2619, but before restarting Zotero Standalone : "db-after_custom_install-before_restart.json" - 9.0 kB
It is deleted when Zotero Standalone is restarted after the installation (no more file to send you obviously but this ID: 5Q3XDMSW

**Comment by retorquere** [comment]
I have a version lined up with logs to the console, but my build service (circleci) has an [outage](http://status.circleci.com/), I must wait for that to clear.

**Comment by retorquere** [comment]
Alright, they're back up. The version can be found [here](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.68-circle-2625.xpi). This version logs to the JS console, which should be available from the start vs the Zotero debugging which kicks in a little late for some things.

The console is usually limited for performance reasons, but since it is also a very busy place, messages often get lost before you can get at them. You can (temporarily, please change back later) the number of lines logged by going into `about:config` and upping the `devtools.hud.loglimit.console` limit. Change that to e.g. 20000, stop Zotero, start Zotero, and open the javascript console (you might have to fiddle with the [command line](https://www.zotero.org/support/reporting_problems) to get that done), and filter on `better-bibtex`. I'm looking for any errors, but also specifically for lines that contain `DBStore`.

**Comment by retorquere** [comment]
(in your latest log, the `TypeError: str.split is not a function` errors are gone BTW -- gotta love transient errors)

**Comment by gracile-fr** [comment]
I'm afraid there's no devtools.hud.loglimit.console in Zotero Standalone about:config. I'll try with Z for firefox (I'm using both actually, and btw don't really understand the BBT warning "Zotero is in connector mode –– not saving database").

**Comment by retorquere** [comment]
Nice. Does the console show anything relevant nonetheless?

**Comment by gracile-fr** [comment]
With Z Firefox and circle-2625 : 

<code>
1469050374496	addons.manager	WARN	Exception calling callback: TypeError: statusButton is null (chrome://zutilo/content/firefoxOverlay.js:327:9) JS Stack trace: getStatusPopup@firefoxOverlay.js:327:9 < setupStatusPopup@firefoxOverlay.js:334:25 < ZutiloChrome.firefoxOverlay.firefoxZoteroOverlay@firefoxOverlay.js:63:9 < ZutiloChrome.firefoxOverlay.init/</overlayFunc@firefoxOverlay.js:31:47 < Zutilo.checkZoteroActiveAndCallIf/<@zutilo.jsm:216:17 < safeCall@AddonManager.jsm:179:5 < makeSafe/<@AddonManager.jsm:195:25 < Handler.prototype.process@Promise-backend.js:937:23 < this.PromiseWalker.walkerLoop@Promise-backend.js:816:7 < this.PromiseWalker.scheduleWalkerLoop/<@Promise-backend.js:750:11 < getContentsFromURL@file.js:214:3 < Zotero.BetterBibTeX.getContentsFromURL@zotero-better-bibtex.js:1764:12 < Zotero.BetterBibTeX.load@zotero-better-bibtex.js:1778:12 < Zotero.BetterBibTeX.loadTranslators@zotero-better-bibtex.js:1468:5 < Zotero.BetterBibTeX.init@zotero-better-bibtex.js:991:3 < load@include.js:14:5 < exports.Utils.yield@utils.js:361:12 < INIParser.prototype.process@filterStorage.js:798:7 < exports.IO.readFromFile/onProgress@io.js:97:15 < exports.IO.readFromFile/<@io.js:182:11 < TaskImpl_run@Task.jsm:319:40 < Handler.prototype.process@Promise-backend.js:937:23 < this.PromiseWalker.walkerLoop@Promise-backend.js:816:7 < this.PromiseWalker.scheduleWalkerLoop/<@Promise-backend.js:750:11


invalid 'in' operand loadInfo
PolicyImplementation.asyncOnChannelRedirect()
 contentPolicy.js:380
getContentsFromURL()
 file.js:214
Zotero.BetterBibTeX.getContentsFromURL()
 zotero-better-bibtex.js:1764
Zotero.BetterBibTeX.load()
 zotero-better-bibtex.js:1778
Zotero.BetterBibTeX.loadTranslators()
 zotero-better-bibtex.js:1468
Zotero.BetterBibTeX.init()
 zotero-better-bibtex.js:991
load()
 include.js:14
exports.Utils.yield()
 utils.js:361
INIParser.prototype.process()
 filterStorage.js:798
exports.IO.readFromFile/onProgress()
 io.js:97
exports.IO.readFromFile/<()
 io.js:182
InterpretGeneratorResume()
 self-hosted:1040
next()
 self-hosted:947
TaskImpl_run()
 Task.jsm:319
bound TaskImpl_run()
 self-hosted:792
Handler.prototype.process()
 Promise-backend.js:937
this.PromiseWalker.walkerLoop()
 Promise-backend.js:816
bound ()
 self-hosted:750
bound bound ()
 self-hosted:750
this.PromiseWalker.scheduleWalkerLoop/<()
 Promise-backend.js:750
 contentPolicy.js:380:0
</code>

**Comment by retorquere** [comment]
Oh, eh, the latter warning means zotero is in connector mode. Zotero
triggers connector mode when both zotero Firefox and zotero standalone are
running simultaneously, in which case zotero Firefox behaves like the
connector plug-ins such as for chrome. This is mainly because having both
zoteros (and bbts) acting on the same database is likely to corrupt the
databases.

**Comment by gracile-fr** [comment]
(I understand what is the connector mode ; what I don't get is the BBT behaviour in this case)

**Comment by retorquere** [comment]
Same use-case as zotero itself - having both bbts acting in the same
profile is going to corrupt the bbt database.

**Comment by gracile-fr** [comment]
(ok, but it seems to me that bbt displays this warning even if Z for Firefox is not started ; but I'm not sure and that's another issue).
What do you think of the report above : https://github.com/retorquere/zotero-better-bibtex/issues/537#issuecomment-234093652 ?

**Comment by retorquere** [comment]
I get a notification from Zotero when it changes into connector mode -- when I get that, I follow suit (and back). So if BBT thinks it's in connector mode, it's because Zotero thinks it's in connector mode.

That's a strange error. Can you disable Zutilo temporarily and see if the problem goes away?

**Comment by gracile-fr** [comment]
I start ZStandalone, disable Zutilo, restart, create an auto-export, close ZS, re-start it. The auto-export doesn't appear anymore in the BBT prefs. I close Standalone.
In Firefox, I open Zotero (Zutilo isn't disabled there) and get these two warnings (not once…):
<code>"Citation key rescan after upgrade"</code>
<code>"DBStore : failed to load cache"</code>

I'm using bbt circle 2625 in both Standalone and Firefox

**Comment by gracile-fr** [comment]
Restarting Firefox with Zutilo disabled, the javascript console displays this (for better-bibtex filter):

invalid 'in' operand loadInfo
PolicyImplementation.asyncOnChannelRedirect()
 contentPolicy.js:380
getContentsFromURL()
 file.js:214
Zotero.BetterBibTeX.getContentsFromURL()
 zotero-better-bibtex.js:1764
Zotero.BetterBibTeX.load()
 zotero-better-bibtex.js:1778
Zotero.BetterBibTeX.loadTranslators()
 zotero-better-bibtex.js:1468
Zotero.BetterBibTeX.init()
 zotero-better-bibtex.js:991
load()
 include.js:14
exports.Utils.yield()
 utils.js:361
INIParser.prototype.process()
 filterStorage.js:798
exports.IO.readFromFile/onProgress()
 io.js:97
exports.IO.readFromFile/<()
 io.js:182
InterpretGeneratorResume()
 self-hosted:1040
next()
 self-hosted:947
TaskImpl_run()
 Task.jsm:319
bound TaskImpl_run()
 self-hosted:792
Handler.prototype.process()
 Promise-backend.js:937
this.PromiseWalker.walkerLoop()
 Promise-backend.js:816
bound ()
 self-hosted:750
bound bound ()
 self-hosted:750
this.PromiseWalker.scheduleWalkerLoop/<()
 Promise-backend.js:750
 contentPolicy.js:380:0

**Comment by retorquere** [comment]
BTW, you're on an older version of Zutilo (1.4.0pre001 where the current version is 2.0.2), and I think the error that was in the console log has since been fixed.

The "Citation scan" message is normal if you've installed a new BBT version, and each circle version counts as a new version.

I can't find the message "failed to load cache" in my code anywhere -- the closest I can find is `DBStore: failed to load cache.json`, but that would be followed by a reason for the failure. Was the log line perhaps wrapped?

**Comment by retorquere** [comment]
The reason why I'm looking at Zutilo is that I've seen these kinds of things before -- other extensions running asynchronously (which is the in thing to do with javascript) concurrently (insofar that exists in javascript at all) with BBT calling Zotero.File.getContents, the async code erroring out, and interrupting the in-progress call to Zotero.File.getContents.

If the problem persists even without Zutilo, I'd need the full stack trace that the error above is embedded in.

**Comment by gracile-fr** [comment]
and FWIW, in my zotero data directory\better-bibtex, I have two files: "cache.json.5" and db.json.5" and the "translator" directory. I mean it's not only db.json which has been removed but also cache.json… I hope that helps.

**Comment by gracile-fr** [comment]
Yes it's probably "DBStore: failed to load cache.json", the message was partly hidden.

**Comment by retorquere** [comment]
db.json and cache.json are handled analogously at startup, so what affects the one will affect the other. It's only after startup that their timing starts to diverge.

The message after `DBStore: failed` will give the reason of the failure. One failure could be that the file is missing, which means it's been deleted before the DB loader ever gets to it.

The translator directory is a remnant of older times and can be removed.

**Comment by gracile-fr** [comment]
ok, what should I do then ?
The DBStore: failed message only appears when I switch between Firefox and Standalone.

**Comment by gracile-fr** [comment]
ok, what should I do then ?
The DBStore: failed message only appears when I switch from Standalone to Firefox [edited].
In the jsconsole, I get this:
TypeError: Zotero.BetterBibTeX is undefined zotero-better-bibtex.js:1354:7
TypeError: Zotero.BetterBibTeX is undefined zotero-better-bibtex.js:1333:1

**Comment by retorquere** [comment]
The `undefined` errors mean there's an early startup error preventing BBT from initializing. I'll need a real-time debug output log, but on Windows, [that's hard work to do](https://www.zotero.org/support/debug_output). Without it I don't really know what's going on alas.

**Comment by gracile-fr** [comment]
Zotero Standalone, BBT circle-2625:

<a href="https://gist.github.com/gracile-fr/42ae034473752edcd78ab2099171eb79">1st log</a>: start standalone, create an auto-export, close standalone

<a href="https://gist.github.com/gracile-fr/76d3c195aa2e391088b539c91a4b76a7">2nd log</a>: start standalone, create another auto-export (as it seems that the previous one has not been saved…), add new item to the auto-exported collection, then delete it from the collection (should trigger two auto-exports), close standalone.

**Comment by retorquere** [comment]
I just found out why the reason as missing from the logs: I was showing it in a popup rather than logging it. New version that corrects that at https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.68-circle-2627.xpi.

**Comment by gracile-fr** [comment]
So what do you want ? Real-time debug with this version ?

**Comment by retorquere** [comment]
Yes please, for this procedure:

* Start Zotero and verify db.json is present after start
* Stop Zotero  and verify db.json is present after stop
* Start zotero with a real-time debug log

**Comment by retorquere** [comment]
Can you try https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.68-circle-2632.xpi ? I think this will fix the issue.

**Comment by gracile-fr** [comment]
Sorry for the delay. Just tested circle-2632: it works perfectly well so far ! Well done and many many thanks ! (But there's no db.json.saving which appears in the better-bibtex directory)

**Comment by retorquere** [comment]
It's renamed before you noticed it. If you look at the file system events it will have been there.

**Comment by retorquere** [comment]
My tests are running again, I'll cut a new release when they're done (approx 20 minutes in the case of a release)

**Comment by retorquere** [comment]
and super thanks for reporting this!

**Comment by retorquere** [comment]
(and the help with debugging -- it was your suggestion that it was the backup mechanism that set me on the right path)

**Comment by retorquere** [comment]
@gracile-fr do you still have Z4 operational? You have a pretty big library I think, can you check the file sizes for the BBT files in the data dir?

**Comment by gracile-fr** [comment]
Yes. I need BBT so I'm stuck to Z4! There are 2259 items in My Library.
My BBT version is probably out of date (1.6.91).

In <code>zotero</code> data directory (no <code>betterbibtex.sqlite</code>, why?):

File | Size | Date
------------|------------ |------------ 
betterbibtex.sqlite.bak | 1 933 312 | 29/10/2015 19:49
betterbibtex-lokijs.sqlite-journal | 197 168 | 15/07/2016 19:53
betterbibtex-lokijs.sqlite.bak | 196 608 | 15/07/2016 16:52
betterbibtex-lokijs.sqlite.failedmigration | 196 608 | 15/07/2016 19:53
betterbibtex-lokijs.sqlite.1.bak | 196 608 | 13/07/2016 19:39
betterbibtex-lokijs.sqlite-1.tmp | 327 680 | 13/05/2016 17:19
betterbibtex-lokijs.sqlite-2.tmp | 753 664 | 26/05/2016 14:59

In <code>better-bibtex</code> subdirectory:

File | Size | Date
------------|------------ |------------ 
cache.json | 712 889 | 09/08/2017 16:19
cache.json.1 | 712 887 | 09/08/2017 16:00
cache.json.2 | 712 889 | 09/08/2017 02:05
cache.json.3 | 712 887 | 08/08/2017 17:55
cache.json.4 | 712 889 | 08/08/2017 17:40
cache.json.5 | 712 887 | 07/08/2017 16:11
db.json | 184 267 | 09/08/2017 16:19
db.json.1 | 184 264 | 09/08/2017 15:59
db.json.2 | 184 267 | 09/08/2017 02:05
db.json.3 | 184 264 | 08/08/2017 17:54
db.json.4 | 184 264 | 08/08/2017 17:54
db.json.5 | 184 267 | 08/08/2017 17:40

**Comment by retorquere** [comment]
A lot of those will be superfluous but I don't recall exactly in which version what went away; for 1.6.100 you only need the json files.

I had hoped you had more references (I recall someone having 20k but I don't know who it was); but an extrapolation would put 20k refs at 7MB for the cache, and 2MB for the database. That's not too bad. I'm contemplating whether I should start saving data in the main DB for BBT5; the main DB takes care of backups, so I wouldn't have to do it myself, but it's async so I need to make sure the DB is ready before anything else happens (which is not as easy as it should be)

Anyhow, thanks, it's still a data point.

**Comment by dbobak** [comment]
@retorquere, I have a portable installation of Zotero 4, so if you need data from a bit bigger library (3101 references), here they are:

```
 Directory of c:\Users\deni\zoteroPortable

2015-06-15  12:50                53 better-bibtex-serialized-items.json
2016-07-11  14:34           754˙360 betterbibtex-lokijs.sqlite-journal
2016-06-24  12:17        13˙664˙256 betterbibtex-lokijs.sqlite.1.bak
2016-07-11  13:02        13˙664˙256 betterbibtex-lokijs.sqlite.bak
2015-03-04  15:59           491˙520 betterbibtex.sqlite.1.bak
2015-10-29  20:47           786˙432 betterbibtex.sqlite.bak

 Directory of c:\Users\deni\zoteroPortable\better-bibtex

2017-08-09  16:54        14˙567˙013 cache.json
2017-08-08  20:24        14˙567˙014 cache.json.1
2017-08-08  20:23        14˙567˙012 cache.json.2
2017-08-07  15:48        14˙614˙384 cache.json.3
2017-08-07  12:54        14˙620˙856 cache.json.4
2017-08-07  12:54        14˙605˙326 cache.json.5
2017-08-09  16:54           643˙955 db.json
2017-08-09  16:53           643˙953 db.json.1
2017-08-08  20:24           643˙955 db.json.2
2017-08-08  20:23           643˙954 db.json.3
2017-08-08  20:22           643˙947 db.json.4
2017-08-07  15:48           636˙471 db.json.5
              12 File(s)     91˙397˙840 bytes
               2 Dir(s)  500˙791˙230˙464 bytes free
```

**Comment by retorquere** [comment]
Is that a 14mb cache from 3.5k references? Looking only at the .json files.

**Comment by dbobak** [comment]
Yes, I don't know why. After resetting the cache, `cache.json.1` is 2,3mb, `cache.json` still 14mb. Entire library, exported do Biblatex (with abstracts) is about 3,4mb.

**Comment by retorquere** [comment]
I'd have to look into that. Still, those aren't numbers I'd want to do deep voodoo for just to shave a little time off saving.
